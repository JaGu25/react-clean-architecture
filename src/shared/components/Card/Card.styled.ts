import styled from 'styled-components'
import { boxShadow, colors, fontSizes, md, spacing } from '../../utils/styles.utils'

const CardWrapper = styled.div`
    border-radius: ${spacing.Spacing1};
    box-shadow: ${boxShadow.basic};
    display: flex;
    flex-direction: column;
    width: 100%;
    transition: all 0.2s ease-in-out;

    :hover {
        transform: scale(1.05);
    }
`

const CardContent = styled.div`
    display: flex;
    flex-direction: column;
    height: 100%;
    justify-content: space-between;
    padding: ${spacing.Spacing2};
`

const CardImage = styled.img`
    max-width: 100%;
    height: 250px;
    object-fit: fill;
`

const CardTitle = styled.h2`
   font-size: ${fontSizes.lg};
   font-weight: 700;
   color: ${colors.text};
`

const CardDescription = styled.p`
   margin: ${spacing.Spacing1} ${spacing.Spacing0} ${spacing.Spacing4};
   font-size: ${fontSizes.sm};
   font-weight: 400;
   color: ${colors.text};
`

const CardActions = styled.div`
    display: flex;
    justify-content: space-between;
    align-items: center;
`

const CardLikeIcon = styled.i`
    font-size: ${fontSizes.lg};
    cursor: pointer;
    &.active {
        color: ${colors.red};
    }
`

const CardStyled = {
    CardWrapper,
    CardImage,
    CardContent,
    CardTitle,
    CardDescription,
    CardActions,
    CardLikeIcon
}

export default CardStyled