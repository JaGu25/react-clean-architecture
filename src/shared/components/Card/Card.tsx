import React, { useState } from 'react'
import CardStyled from './Card.styled'
import Button from '../Button/Button'

interface IProps {
    description: string
    title: string
    img: string,
    handleClick: () => void
    onLikedChange: (isLiked: boolean) => void
    isLiked?: boolean
}

const Card: React.FC<IProps> = ({
    description,
    img,
    title,
    handleClick,
    onLikedChange,
    isLiked = true
}) => {

    const [liked, setLiked] = useState(isLiked);
    const handleLikeChange = (): void => {
        setLiked(!liked)
        onLikedChange(!liked)
    }

    return (
        <CardStyled.CardWrapper>
            <CardStyled.CardImage src={img} />
            <CardStyled.CardContent>
                <div>
                    <CardStyled.CardTitle>
                        {title}
                    </CardStyled.CardTitle>
                    <CardStyled.CardDescription>
                        {description}
                    </CardStyled.CardDescription>
                </div>
                <CardStyled.CardActions>
                    <Button
                        text='Ver mÃ¡s'
                        type='primary'
                        onClick={handleClick} />
                    <CardStyled.CardLikeIcon
                        data-testid="likeaction"
                        className={`${liked ? 'fa-solid ' : 'fa-regular'} fa-heart ${liked && ('active')}`}
                        onClick={handleLikeChange}
                    />
                </CardStyled.CardActions>
            </CardStyled.CardContent>
        </CardStyled.CardWrapper>
    )
}

export default Card
