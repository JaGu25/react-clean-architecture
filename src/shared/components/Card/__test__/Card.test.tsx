import React from "react"
import { fireEvent, render, screen } from "@testing-library/react"
import Card from "../Card"

const click = jest.fn()
const onLikedChange = jest.fn()

describe('Card Component', () => {
    it('should render without crash', () => {
        const { container } = render(
            <Card
                title="titulo"
                description="description"
                handleClick={() => { }}
                onLikedChange={() => { }}
                img=""
            />
        )
        expect(container).toMatchSnapshot()
    })

    it('should click button', async () => {
        render(
            <Card
                title="titulo"
                description="description"
                handleClick={click}
                onLikedChange={() => { }}
                img=""
            />
        )

        const button = await screen.findByText('Ver mÃ¡s')
        fireEvent.click(button)
        expect(click).toBeCalled()
    })

    it('should listen if like icon is clicked', async () => {
        render(
            <Card
                title="titulo"
                description="description"
                handleClick={click}
                onLikedChange={onLikedChange}
                img=""
            />
        )

        const likeButton = await screen.findByTestId('likeaction')
        fireEvent.click(likeButton)
        expect(onLikedChange).toBeCalled()
    })
})