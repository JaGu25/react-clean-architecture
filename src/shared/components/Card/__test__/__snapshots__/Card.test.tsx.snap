// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component should render without crash 1`] = `
<div>
  <div
    class="sc-bczRLJ gpCSHu"
  >
    <img
      class="sc-dkzDqf bxnkfQ"
      src=""
    />
    <div
      class="sc-gsnTZi gzgePQ"
    >
      <h2
        class="sc-hKMtZM lptbku"
      >
        titulo
      </h2>
      <p
        class="sc-eCYdqJ ehJxse"
      >
        description
      </p>
      <div
        class="sc-jSMfEi NutLz"
      >
        <button
          class="sc-iBkjds duwzBz"
          color="#6200EE"
          type="button"
        >
          Ver m√°s
        </button>
        <i
          class="sc-gKXOVf kQPepr fa-solid  fa-heart active"
          data-testid="likeaction"
        />
      </div>
    </div>
  </div>
</div>
`;
